src/rfc2047.rs

```diff
@@ -40,38 +39,34 @@ pub fn decode_rfc2047(s: &str) -> Option<String> {
     }
 }
 
-#[stable]
 pub fn decode_q_encoding(s: &str) -> Result<Vec<u8>, String> {
     let mut result = Vec::new();
+    let mut char_iter = s.chars();
 
-    let mut pos = 0;
-
-    while pos < s.len() {
-        let c = s.char_range_at(pos);
-        pos = match c.ch {
-            '=' => {
-                let mut inner_pos = c.next;
+    loop {
+        match char_iter.next() {
+            Some('=') => {
                 let mut hex_string = String::new();
                 for _ in 0..2 {
-                    let hex_digit_char = s.char_range_at(inner_pos);
-                    hex_string.push(hex_digit_char.ch);
-                    inner_pos = hex_digit_char.next;
-                }
+                    let hex_digit_char = char_iter.next().unwrap();  // FIXME
```

Comment:
I find this highly problematic but am unsure if and how to fix. I have no idea how the previous code even behaved if it hit string boundaries?